### ISSUE TITLE
Model Training Experiments

### ISSUE THREAD

**Issue by Tandon-A** [Healthcare-Robotics/BodyPressure#7]
Hello Henry, 

This is fantastic work. Thank you for providing the code base and the BodyPressureSD dataset. 

I had a few questions related to the modeling experiments. 

1. I understand that all models were trained separately and then fixed while training the next model. E.g.- BetaNet is trained once and fixed. Similarly, Mod1 is fixed to train Mod2. 
a) Did you perform any end-to-end training experiments? If yes, what were the metrics? If not, what were the issues?
b) Did you perform any experiments where the earlier model is not fixed, but fine-tuned with a lower learning rate? 

2. When training Mod1 you do not apply Lv2v1 (L2 on vertices)? I guess since Mod1 is giving only an initial pose, it is okay if it is not perfect, but wouldn't using this loss help Mod1 give a better estimate? 

3. You mention that no loss on the output of DMR from Mod1 is applied. Any reasons for this? 
a) Did you perform any experiments applying loss on estimated DMR output and GT DMR output in Mod1? 
b) Did you perform any experiments using DMR in Mod2 and applying loss on DMR output to help the model learn better?

4.  In the paper you mentioned that you used the same hyperparameters for training BPBNet and BPWNet. I wasn't able to understand the reason behind this. Isn't it possible that the best hyperparameters for BPBNet could result in BPBNet performing better?

5. Any experiments on using a pre-trained ResNet as the encoder? I understand that depth is just 1 channel, but maybe by copying it to all three channels?

Best, 
Abhishek

**Comment by henryclever** [comment]
Hi Abhishek,

These are very good questions! Thank you.

1. (a) I did not perform and end to end training. It's possible now that I could since I have an A6000 GPU but I did not at that time, so there were issues fitting things on my machine without having to significantly decrease batch size. A big limitation I ran into is that I had to make repeats of all the SMPL torch params for each item in the batch - see lines 239-276 here: https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/mesh_depth_lib_bp.py . So for a smaller GPU you have to break this computation into sub batches and run them through a for loop, which slows things down. If the capabilities of that torch version were such that I could do all the matmuls without making these repeats, it would have a lower memory footprint and increase the chances of fitting everything on less VRAM. Maybe torch can do that now, I just don't know. Future work! (b) no, I did not.
2. It might, but my guess was that this amount of improvement would be relatively minor. I remember training Mod2 without the vertex loss and the difference was marginal. So if Mod1 was supposed to only provide a rough estimate, it seemed to me that it would not matter so much. Furthermore, adding the v2v loss to Mod1 adds the whole SMPL model to backprop which makes it more than 2x slower to train from my memory, so I just went with the faster solution.
3. (a) Same reason as (2). (b) I did not -- it's possible this could improve the results, I just didn't try it.
4. Yes, that's possible, but I didn't spend a great amount of time tuning the hyperparameters anyway. So I felt I was not over-optimizing for either network; rather I just picked values I saw in previous papers and had those I used in the past for other projects that seemed to work well. My guess is the results could be improved with better networks; I was more interested in the new ideas and concepts than getting my numbers up 5% or w/e.
5. I do remember trying a pretrained ResNet in an early iteration of the work with a different dataset, but I remember it not helping that much so I shelved it and didn't try again. But it's possible a pretrained ResNet with depth inputs to all 3 RGB channels could help performance with the current dataset. One reason I didn't try this is because my dataset is fairly large - 100K examples - so my thinking was it would forget most of the stuff it had been pretrained on, but i could be wrong.

-Henry

**Comment by Tandon-A** [comment]
Thanks for the info!
